packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -1,242 +1,256 @@
+        }
+      },
+    );
 
-      if (isDetectionEnabledForNetwork && isChainIdChanged) {
-        this.detectTokens();
-      }
-    });
+    onPreferencesStateChange(
+      async ({ selectedAddress: newSelectedAddress, useTokenDetection }) => {
+        const isSelectedAddressChanged =
+          this.#selectedAddress !== newSelectedAddress;
+        const isDetectionChangedFromPreferences =
+          this.#isDetectionEnabledFromPreferences !== useTokenDetection;
+
+        this.#selectedAddress = newSelectedAddress;
+        this.#isDetectionEnabledFromPreferences = useTokenDetection;
+
+        if (
+          useTokenDetection &&
+          (isSelectedAddressChanged || isDetectionChangedFromPreferences)
+        ) {
+          await this.detectTokens();
+        }
+      },
+    );
+
+    this.messagingSystem.subscribe(
+      'NetworkController:networkDidChange',
+      async ({ providerConfig: { chainId: newChainId } }) => {
+        this.#isDetectionEnabledForNetwork =
+          isTokenDetectionSupportedForNetwork(newChainId);
+        const isChainIdChanged = this.#chainId !== newChainId;
+
+        this.#chainId = newChainId;
+
+        if (this.#isDetectionEnabledForNetwork && isChainIdChanged) {
+          await this.detectTokens();
+        }
+      },
+    );
   }
 
   /**
    * Start polling for detected tokens.
    */
   async start() {
-    this.configure({ disabled: false });
-    await this.startPolling();
+    await this.#startPolling();
   }
 
   /**
    * Stop polling for detected tokens.
    */
   stop() {
-    this.configure({ disabled: true });
-    this.stopPolling();
+    this.#stopPolling();
   }
 
-  private stopPolling() {
-    if (this.intervalId) {
-      clearInterval(this.intervalId);
+  #stopPolling() {
+    if (this.#intervalId) {
+      clearInterval(this.#intervalId);
     }
   }
 
   /**
    * Starts a new polling interval.
-   *
-   * @param interval - An interval on which to poll.
    */
-  private async startPolling(interval?: number): Promise<void> {
-    interval && this.configure({ interval }, false, false);
-    this.stopPolling();
+  async #startPolling(): Promise<void> {
+    this.#stopPolling();
     await this.detectTokens();
-    this.intervalId = setInterval(async () => {
+    this.#intervalId = setInterval(async () => {
       await this.detectTokens();
-    }, this.config.interval);
+    }, this.getIntervalLength());
   }
 
-  private getCorrectChainId(networkClientId?: NetworkClientId) {
-    if (networkClientId) {
-      return this.getNetworkClientById(networkClientId).configuration.chainId;
+  #getCorrectChainId(networkClientId?: NetworkClientId) {
+    const { chainId } =
+      this.messagingSystem.call(
+        'NetworkController:getNetworkConfigurationByNetworkClientId',
+        networkClientId ?? this.#chainId,
+      ) ?? {};
+    if (chainId) {
+      this.#chainId = chainId;
     }
-    return this.config.chainId;
+    return this.#chainId;
```

Comment:
Hmm... a network client ID isn't the same as a chain ID. So I'm not sure that this will work. We might have to do something a bit more verbose:

``` suggestion
    let chainId: Hex;

    if (networkClientId) {
      const networkConfiguration = this.messagingSystem.call(
        'NetworkController:getNetworkConfigurationByNetworkClientId',
        networkClientId,
      );
      chainId = networkConfiguration?.chainId;
    }

    chainId ??= this.#chainId;

    return chainId;
```

Reply:
This reverts the `should detect tokens correctly on the Polygon network` test to the previous breaking state. 

It looks like line 241 `networkClientId ?? this.#chainId,` is necessary for `detectTokens()` to run with the new chainId updated by the `changeNetwork()` call assigned to its `networkClientId` argument. 

`this.#chainId` is the only source of information that's updated by `changeNetwork()`, so if we can't use that, then the user will need to explicitly pass in the new `networkClientId` into a `detectTokens()` call, meaning `.start()` can't be used after a network change.

Reply:
Ah, right.

I see that when we are listening for NetworkController state changes, we grab the chain ID from the provider config. In addition to `providerConfig`, there is also a `selectedNetworkClientId` property on the NetworkController state, and if we're using network client ID here, then I think we ought to grab that and save that as a private property instead of chain ID. Then we can use `networkClientId ?? this.#networkClientId` here.

Would that work?

Reply:
Nice, this works! Applied here: https://github.com/MetaMask/core/pull/3609/commits/a41a218e9c84b1bf337cebf7283889c56d3774cb

Reply:
Nice, this works! Applied here: https://github.com/MetaMask/core/pull/3609/commits/d32cdf43d616fab43a50aa140282a7a66dffc9bd